idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,230 @@
+        theta_names(list of strings)   : List of estimated Var names
+        var_dic(dictionary)            : If any original variable contains apostrophes, need an auxiliary dictionary with keys theta_namess without apostrophes, values with apostrophes. e.g) var_dic: {'fs.properties.tau[benzene,toluene]': "fs.properties.tau['benzene','toluene']", 'fs.properties.tau[toluene,benzene]': "fs.properties.tau['toluene','benzene']"} 
+        tee(bool, optional)            : Indicates that ef solver output should be teed
+        solver_options(dict, optional) : Provides options to the solver (also the name of an attribute)
+
+    Returns:
+            : It returns the following variables:
+            
+            - gradient_f_dic(dic)            : Gradient of the objective function with respect to the (decision variables, parameters) at the optimal solution with variable name as key e.g) dic = {d(f)/d(x1):0.1, d(f)/d(x2):0.1}
+            - gradient_c_dic(dic)            : Gradient of the constraints with respect to the (decision variables, parameters) at the optimal solution with constraint number and variable name as key e.g) dic = {d(c1)/d(x1):1.1, d(c4)/d(x2):0.1}. Only non-zero gradients are included.
+            - dsdp_dic(dict)                 : Gradient vector of the (decision variables, parameters) with respect to paramerters (=theta_name). e.g) dict = {'d(x1)/d(p1)': 1.0, 'd(x2)/d(p1)': 0.0, 'd(p1)/d(p1)': 1.0, 'd(p2)/d(p1)': 0.0, 'd(x1)/d(p2)': 0.0, 'd(x2)/d(p2)': 1.0, 'd(p1)/d(p2)': 0.0, 'd(p2)/d(p2)': 1.0}
+            - propagation_f(dictionary)      : Error propagation in the objective function with the dictionary key, 'objective', df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+            - propagation_c(dictionary)      : Error propagation in the constraints with the dictionary keys, 'constraints r' where r is the line number. If no constraint includes uncertain parameters, return an empty dictionary      
+
+    Raises:
+        Exception:  When model_uncertain is neither 'ConcreteModel' nor 'function'.
+    """
+
+    # define Pyomo model
+    if type(model_uncertain).__name__ == 'ConcreteModel':
+        model = model_uncertain
+    elif type(model_uncertain).__name__ == 'function':
+        model = model_uncertain()
+    else:
+        raise TypeError('model_uncertain must be either python function or Pyomo ConcreteModel.')
+    # Remove all "'" in theta_names     
+    theta_names, var_dic,variable_clean = clean_variable_name(theta_names)
+    for v in theta_names:
+        eval('model.'+var_dic[v]).setlb(theta[v])
+        eval('model.'+var_dic[v]).setub(theta[v])
+    # get gradient of the objective function, constraints, and the column number of each theta
+    gradient_f,gradient_f_dic, gradient_c,gradient_c_dic, line_dic = get_dfds_dcds(model, theta_names, tee)
+    dsdp_dic, col  = get_dsdp(model, theta_names, theta, var_dic,tee)      
+    x_list = [x for x in col if x not in var_dic.keys()]
```

Comment:
Add more comments please. What is going on here?

Reply:
`x_list` includes only decision variables (exclude theta_names). `col` includes both decision variables and `parameters`. I added a comment.